soroban-env-host/src/host/error.rs

```diff
@@ -229,40 +214,48 @@ impl Host {
-            // log, failing gracefully (just emitting a no-debug-info
-            // `HostError` wrapping the supplied `Error`) if we cannot acquire
-            // the refcell. This is to handle the "double fault" case where we
-            // get an error _while performing_ any of the steps needed to record
-            // an error as an event, below.
-            if let Ok(mut events_refmut) = self.0.events.try_borrow_mut() {
-                if let Err(e) = self.err_diagnostics(events_refmut.deref_mut(), error, msg, args) {
-                    return e;
+        self.with_debug_budget(
+            || {
+                // We _try_ to take a mutable borrow of the events buffer refcell
+                // while building up the event we're going to emit into the events
+                // log, failing gracefully (just emitting a no-debug-info
+                // `HostError` wrapping the supplied `Error`) if we cannot acquire
+                // the refcell. This is to handle the "double fault" case where we
+                // get an error _while performing_ any of the steps needed to record
+                // an error as an event, below.
+                // We do the event-borrowing ourselves here rather than calling
+                // self.try_borrow_events_mut because we can/should only be called
+                // with an already-borrowed events buffer (to insulate against
+                // double-faulting).
+                if let Ok(mut events_refmut) = self.0.events.try_borrow_mut() {
+                    self.record_err_diagnostics(events_refmut.deref_mut(), error, msg, args);
                 }
-            }
-            let info = self.maybe_get_debug_info();
-            return HostError { error, info };
-        }
-        error.into()
+                Ok(HostError {
+                    error,
+                    info: self.maybe_get_debug_info(),
+                })
+            },
+            || error.into(),
+        )
     }
 
     pub(crate) fn maybe_get_debug_info(&self) -> Option<Box<DebugInfo>> {
-        if let Ok(true) = self.is_debug() {
-            if let Ok(events_ref) = self.0.events.try_borrow() {
-                let events = match self
-                    .as_budget()
-                    .with_free_budget(|| events_ref.externalize(self))
-                {
-                    Ok(events) => events,
-                    Err(_) => return None,
-                };
-                let backtrace = Backtrace::new_unresolved();
-                return Some(Box::new(DebugInfo { backtrace, events }));
-            }
-        }
-        None
+        self.with_debug_budget(
+            || {
+                // We do the event-borrowing ourselves here rather than calling
+                // self.try_borrow_events because we can/should only be called
+                // with an already-borrowed events buffer (to insulate against
+                // double-faulting).
```

Comment:
This comment is copied from elsewhere and doesn't describe what's happening here; and at this point we also have a fallible helper `Host::get_events` that does the `externalize` call itself, I think we might as well use that and just do an `if let Ok(events) = self.get_events() { ... }` 

(I've also filed https://github.com/stellar/rs-soroban-env/issues/1132 to enrich this path a bit, DebugInfo is a bit too easily lost currently)

Reply:
Removed this comment. 
Sorry about the copy-paste, I originally got tripped on double fault here when I tried to do a cleanup and wrote a long comment, but then saw our comment elsewhere which is better written, so I decided to copy and modify. But realizing now it isn't accurate either. 
(Same as below)